### DISCUSSION TITLE
dovi_tool ported to C, added to ffmpeg hevc decoding

### DISCUSSION THREAD

**Discussion by docdude** [quietvoid/dovi_tool#66] [Ideas]
Thank you for your work. I've managed to port over your code to C and added it to FFMPEG, hevcdec.c and can parse the unspec62 NAL.  I'm using a hwaccel version for Raspberry Pi to see if I can pass the RPU data.  It's WIP but I've got the all of the RPU data parsed.  My main issues currently are difficulty with dynamic arrays in rpu_mapping and rpu_nlq and in the st2094_10 metadata blocks.  Getting C to enum structures is not possible, nor are dynamic structures.  How to pass the buffer containing the parsed data in FFMPEG is another question I have.  I'm not sure at this point if this will lead to anything fruitful, not sure if code Ive added to allow me to create side_data as such will pass on the decoded rpu_data.

`av_frame_new_side_data_from_buf(out, AV_FRAME_DATA_DOVI_RPU, dovi_buf))`

**Comment by quietvoid** [comment]
I don't know what you're asking, and you're not linking to any code.
Are you aware of this patch set? https://patchwork.ffmpeg.org/project/ffmpeg/cover/20211101203227.1648805-1-derek.buitenhuis@gmail.com/
It does what you're trying to do except the parsing of the RPU.

There's a WIP for a C API being developed here: https://github.com/quietvoid/dovi_tool/pull/63
It's currently incomplete because nested `Vec`s are indeed complicated to conveniently send to C through FFI.

**Reply by docdude** [reply]
I wasn't aware of that patch but thank you, it looks like I was adding the RPU to the  frame side data at the wrong time.   I'll contact the author of the patch and send him the ported parsing code. 

If you figure out a way to handle dynamic arrays and/or dynamic nested structures please keep me in the loop.  I'll do the same.

**Reply by quietvoid** [reply]
There's some additional discussion here: https://patchwork.ffmpeg.org/project/ffmpeg/patch/20211021134843.1436241-2-derek.buitenhuis@gmail.com/

**Reply by docdude** [reply]
My current workaround is to set one dimension and multidimensional arrays statically, parse the data and rewrite it pulling the valid bits from each array and pass into a temp buffer only the parts of the array that actually contain data. 

I've contacted the patch author.  Thanks for the lead.

**Reply by quietvoid** [reply]
You're probably better off using `malloc` and creating the dynamic buffers.
An example of how I decided to handle it: https://github.com/quietvoid/dovi_tool/blob/capi_rework/dolby_vision/examples/capi.c
